java/memory/memory-core/src/main/java/org/apache/arrow/memory/util/SingletonAllocator.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.memory.util;
+
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+
+import org.apache.arrow.memory.BufferAllocator;
+import org.apache.arrow.memory.RootAllocator;
+import org.apache.arrow.util.VisibleForTesting;
+
+/**
+ * Factory class for managing allocators in simple applications that make
+ * use of an application-wide singleton root allocator with no listeners,
+ * automatic allocator defaults, and no total allocator limit.
+ * <p>
+ * Child allocators are created with a default name with the format
+ * <code>SingletonAllocator#(AtomicCounter)</code>.
+ */
+public final class SingletonAllocator {
```

Comment:
We need to ban this from usage in Arrow itself. I'm also not convinced this is worth shipping as part of Arrow, vs improving documentation on how allocators are supposed to be used.

Reply:
ohhh ok, then, Does this PR should be closed to add all of this as a proper documentation?

Reply:
@lidavidm I agree that the usage should be banned in `src/main`, but the use of GlobalAllocator in third-party test scripts has uncovered quite a few allocation bugs in Arrow.

We may want to either allow/encourage/mandate the use of GlobalAllocator in the test suite, or provide an additional specialized allocator or factory method for use in the test suite that provides the bug-detecting effects.

Reply:
Yes, the test suite should be closing allocators properly - @davisusanibar can you file an issue to audit the tests? (Possibly, we should define and use a base test class or rule or whatever the JUnit 5 equivalent is.)

Reply:
Just filed this issue https://github.com/apache/arrow/issues/37216, please add suggestion/changes as needed.